scipy/spatial/distance.py

```diff
@@ -1452,51 +1508,44 @@ def dfun(u, v):
-            dm = pdist(X, cosine)
-        elif metric == 'test_correlation':
-            dm = pdist(X, correlation)
-        elif metric == 'test_hamming':
-            dm = pdist(X, hamming)
-        elif metric == 'test_jaccard':
-            dm = pdist(X, jaccard)
-        elif metric == 'test_chebyshev' or metric == 'test_chebychev':
-            dm = pdist(X, chebyshev)
-        elif metric == 'test_yule':
-            dm = pdist(X, yule)
-        elif metric == 'test_matching':
-            dm = pdist(X, matching)
-        elif metric == 'test_dice':
-            dm = pdist(X, dice)
-        elif metric == 'test_kulsinski':
-            dm = pdist(X, kulsinski)
-        elif metric == 'test_rogerstanimoto':
-            dm = pdist(X, rogerstanimoto)
-        elif metric == 'test_russellrao':
-            dm = pdist(X, russellrao)
-        elif metric == 'test_sokalsneath':
-            dm = pdist(X, sokalsneath)
-        elif metric == 'test_sokalmichener':
-            dm = pdist(X, sokalmichener)
+        elif mstr.startswith("test_"):
+            test_metrics = {'test_euclidean': euclidean,
+                            'test_braycurtis': braycurtis,
+                            'test_canberra': canberra,
+                            'test_cityblock': cityblock,
+                            'test_minkowski': minkowski,
+                            'test_wminkowski': wminkowski,
+                            'test_cosine': cosine,
+                            'test_correlation': correlation,
+                            'test_hamming': hamming,
+                            'test_jaccard': jaccard,
+                            'test_chebyshev': chebyshev,
+                            'test_chebychev': chebyshev,
+                            'test_yule': yule,
+                            'test_matching': hamming,
+                            'test_dice': dice,
+                            'test_kulsinski': kulsinski,
+                            'test_rogerstanimoto': rogerstanimoto,
+                            'test_russellrao': russellrao,
+                            'test_sokalsneath': sokalsneath,
+                            'test_sokalmichener': sokalmichener,
+                            'test_sqeuclidean': sqeuclidean,
+                            'test_seuclidean': seuclidean,
+                            'test_mahalanobis': mahalanobis,
+                            }
+            if mstr in test_metrics:
+                dfun = test_metrics[mstr]
+                if p != 2:  # if not default
+                    dfun = partial(dfun, p=p)
+                if w is not None:
+                    dfun = partial(dfun, w=w)
+                elif V is not None:
+                    dfun = partial(dfun, V=V)
+                elif VI is not None:
+                    dfun = partial(dfun, VI=VI)
+                dm = pdist(X, dfun)
+            else:
+                raise ValueError('Unknown Distance Metric: %s' % mstr)
```

Comment:
"test_" versions are the same thing as the python version, the difference is that "test_" lets you use strings to specify in cdist and pdist, which can be nicer than trying to find the function directly. It's mainly used in the testing modules, and to debug (since C gives terrible stack traces).